packages/code-infra/src/cli/cmdPublishCanary.mjs

```diff
@@ -8,26 +8,306 @@
- * @property {boolean} [dryRun] - Whether to run in dry-run mode
- */
-
+import { fetchCommitsBetweenRefs } from '../utils/changelog.mjs';
 import {
-  getWorkspacePackages,
+  getCurrentGitSha,
   getPackageVersionInfo,
+  getWorkspacePackages,
   publishPackages,
   readPackageJson,
-  writePackageJson,
-  getCurrentGitSha,
   semverMax,
+  writePackageJson,
 } from '../utils/pnpm.mjs';
 
+/**
+ * @typedef {Object} Args
+ * @property {boolean} [dryRun] - Whether to run in dry-run mode
+ * @property {boolean} [githubRelease] - Whether to create GitHub releases for canary packages
+ */
+
 const CANARY_TAG = 'canary';
 
+/**
+ * Get Octokit instance with authentication
+ * @returns {Octokit} Authenticated Octokit instance
+ */
+function getOctokit() {
+  return new Octokit({ authStrategy: createActionAuth });
+}
+
+/**
+ * Get current repository info from git remote
+ * @returns {Promise<{owner: string, repo: string}>} Repository owner and name
+ */
+async function getRepositoryInfo() {
+  try {
+    const result = await $`git remote get-url origin`;
+    const url = result.stdout.trim();
+
+    const parsed = gitUrlParse(url);
+    if (parsed.source !== 'github.com') {
+      throw new Error('Repository is not hosted on GitHub');
+    }
+
+    return {
+      owner: parsed.owner,
+      repo: parsed.name,
+    };
+  } catch (/** @type {any} */ error) {
+    throw new Error(`Failed to get repository info: ${error.message}`);
+  }
+}
+
+/**
+ * Extract package name from npm package name for label matching
+ * @param {string} npmPackageName - npm package name (e.g., '@mui/internal-code-infra')
+ * @returns {string} Package name for label (e.g., 'code-infra')
+ */
+function extractPackageNameForLabel(npmPackageName) {
+  // For scoped packages like @mui/internal-code-infra, extract the last part after 'internal-'
+  if (npmPackageName.startsWith('@')) {
+    const parts = npmPackageName.split('/');
+    if (parts.length >= 2) {
+      const packageName = parts[1];
+      // Remove 'internal-' prefix if present
+      return packageName.replace(/^internal-/, '');
+    }
+  }
+  return npmPackageName;
+}
+
+/**
+ * Get merged PRs since the last canary tag
+ * @param {string} owner - Repository owner
+ * @param {string} repo - Repository name
+ * @returns {Promise<Array<import('../utils/changelog.mjs').FetchedCommitDetails>>} List of merged PRs
+ */
+async function getMergedPRsSinceTag(owner, repo) {
+  const octokit = getOctokit();
+  const changelogs = await fetchCommitsBetweenRefs({
+    octokit,
+    lastRelease: CANARY_TAG,
+    release: 'master',
+    repo,
+    org: owner,
+  });
+  return changelogs;
+}
+
+/**
+ * Generate changelog for a package based on PRs
+ * @param {string} packageName - Package name (e.g., 'code-infra')
+ * @param {Array<import('../utils/changelog.mjs').FetchedCommitDetails>} allPRs - All merged PRs
+ * @returns {string} Generated changelog content
+ */
+function generateChangelogForPackage(packageName, allPRs) {
```

Comment:
Wouldn't it be better to filter commits for the changelog based on the files they touch?

Also the "breaking" detection stuff seems unnecessary for our workflow?

Reply:
If we label our PRs correctly with `breaking change` label, it'll make it clear in the changelogs around versions that should not be blindly updated. Its mainly aesthetics for the changelog content.